elisp/gen-autoloads.el

```diff
@@ -0,0 +1,80 @@
+;;; gen-autoloads.el --- generate autoloads file     -*- lexical-binding: t; -*-
+
+;; Copyright 2021 Google LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;     https://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+
+;;; Commentary:
+
+;; Generate an autoloads file.
+;;
+;; Usage:
+;;
+;;   emacs --quick --batch --load=gen-autoloads.el \
+;;     --funcall=elisp/gen-autoloads-and-exit DEST PKGNAME SOURCE...
+;;
+;; Generates an autoloads file DEST for the SOURCE Emacs Lisp files.
+;;
+;; Exits with a zero status only if successful.
+
+;;; Code:
+
+(require 'cl-lib)
+(require 'nadvice)
+(require 'package)
+
+(defun elisp/gen-autoloads-and-exit ()
+  "Generate an autoloads file and exit Emacs.
+See the file commentary for details."
+  (unless noninteractive
+    (error "This function works only in batch mode"))
+
+  ;; Make output deterministic.
+  ;; The autoload definitions file contains the current time, therefore we
+  ;; override `current-time'.
+  (advice-add 'current-time :override (lambda () '(0 0 0 0)))
+  ;; Return constant file attributes to make builds deterministic.
+  (advice-add 'file-attributes :around
+              (lambda (func &rest args)
+                (let ((ret (apply func args)))
+                  (when ret
+                    (cl-destructuring-bind
+                        (type links uid gid atime mtime ctime size mode change inode dev)
+                        ret
+                      (list type links 0 0 '(0 0 0 0) '(0 0 0 0) '(0 0 0 0)
+                            size mode nil 42 42))))))
+  (add-to-list 'ignored-local-variables 'generated-autoload-file)
+
+  (let* ((out (pop command-line-args-left))
+         (pkgname (pop command-line-args-left))
+         (srcs command-line-args-left)
+         (workdir (file-name-as-directory (make-temp-file "workdir" :dir)))
+         ;; Leaving these enabled leads to undefined behavior and doesn’t make
+         ;; sense in batch mode.
+         (attempt-stack-overflow-recovery nil)
+         (attempt-orderly-shutdown-on-fatal-signal nil)
+         (create-lockfiles nil))
+    (unwind-protect
```

Comment:
It might be useful to only delete the work directory in the success case, then users can more easily debug errors with `--sandbox_debug`. (Bazel should clean up temporary action directories anyway.)

Reply:
In that case, we may as well not delete it ever in case someone needs to debug an unexpected success?

Reply:
done